src/main/java/software/amazon/sns/AmazonSNSExtendedAsyncClient.java

```diff
@@ -0,0 +1,185 @@
+import static software.amazon.sns.AmazonSNSExtendedClientUtil.isTotalMessageSizeLargerThanThreshold;
+import static software.amazon.sns.SNSExtendedClientConstants.MULTIPLE_PROTOCOL_MESSAGE_STRUCTURE;
+import static software.amazon.sns.SNSExtendedClientConstants.USER_AGENT_HEADER_NAME;
+
+public class AmazonSNSExtendedAsyncClient extends AmazonSNSExtendedAsyncClientBase implements SnsAsyncClient {
+
+    static final String USER_AGENT_HEADER = Util.getUserAgentHeader(AmazonSNSExtendedAsyncClient.class.getSimpleName());
+    private static final Log LOGGER = LogFactory.getLog(AmazonSNSExtendedAsyncClient.class);
+    private SNSExtendedAsyncClientConfiguration snsExtendedClientConfiguration;
+    private PayloadStoreAsync payloadStore;
+
+    /**
+     * Constructs a new Amazon SNS extended client to invoke service methods on
+     * Amazon SNS with extended functionality using the specified Amazon SNS
+     * client object.
+     *
+     * <p>
+     * All service calls made using this client are asynchronous, and will return
+     * immediately with a {@link CompletableFuture} that completes when the operation
+     * completes or when an exception is thrown. Argument validation exception are thrown
+     * immediately, and not through the future.
+     * </p>
+     *
+     * @param snsClient
+     *              The Amazon SNS client to use to connect to Amazon SNS.
+     */
+    public AmazonSNSExtendedAsyncClient(SnsAsyncClient snsClient) {
+        this(snsClient, new SNSExtendedAsyncClientConfiguration());
+    }
+
+    /**
+     * Constructs a new Amazon SNS extended client to invoke service methods on
+     * Amazon SNS with extended functionality using the specified Amazon SNS
+     * client object.
+     *
+     * <p>
+     * All service calls made using this client are asynchronous, and will return
+     * immediately with a {@link CompletableFuture} that completes when the operation
+     * completes or when an exception is thrown. Argument validation exceptions are thrown
+     * immediately, and not through the future.
+     * </p>
+     *
+     * @param snsClient
+     *            The Amazon SNS async client to use to connect to Amazon SNS.
+     * @param clientConfig
+     *            The extended client configuration options controlling the
+     *            functionality of this client.
+     */
+    public AmazonSNSExtendedAsyncClient(SnsAsyncClient snsClient, SNSExtendedAsyncClientConfiguration clientConfig) {
+        super(snsClient);
+        this.snsExtendedClientConfiguration = new SNSExtendedAsyncClientConfiguration(clientConfig);
+        S3AsyncDao s3Dao = new S3AsyncDao(snsExtendedClientConfiguration.getS3AsyncClient());
+        this.payloadStore = new S3BackedPayloadStoreAsync(s3Dao, snsExtendedClientConfiguration.getS3BucketName());
+    }
+
+    /**
+     * Constructs a new Amazon SNS extended client to invoke service methods on
+     * Amazon SNS with extended functionality using the specified Amazon SNS
```

Comment:
The class description does not explain how the Async client is different from a normal client. We can consider adding it in the decription.

Reply:
The seconds paragraph did explain the feature of this new client:

```
All service calls made using this client are asynchronous, and will return
     * immediately with a {@link CompletableFuture} that completes when the operation
     * completes or when an exception is thrown. Argument validation exception are thrown
     * immediately, and not through the future.
```